File: src/main/java/xyz/phanta/cbtweaker/buffer/VirtualBufferType.java
package xyz.phanta.cbtweaker.buffer;

import com.google.gson.JsonObject;
import io.github.phantamanta44.libnine.util.Unit;
import io.github.phantamanta44.libnine.util.function.ICapabilityInstanceConsumer;
import mezz.jei.api.IJeiHelpers;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import xyz.phanta.cbtweaker.gui.ScreenRegion;
import xyz.phanta.cbtweaker.integration.jei.render.JeiUiElement;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.Collections;

public abstract class VirtualBufferType<B> implements BufferType<B, Unit, Unit, Unit> {

    @Override
    public Class<Unit> getAccumulatorClass() {
        return Unit.class;
    }

    @Override
    public BufferFactory<B, Unit> loadBufferFactory(JsonObject config) {
        VirtualBufferFactory<B> bufFactory = loadVirtualBufferFactory(config);
        return new BufferFactory<B, Unit>() {
            @Override
            public B createBuffer(World world, BlockPos pos, @Nullable BufferObserver observer) {
                return bufFactory.createBuffer(world, pos, observer);
            }

            @Override
            public Unit createJeiBuffer() {
                return Unit.INSTANCE;
            }
        };
    }

    protected abstract VirtualBufferFactory<B> loadVirtualBufferFactory(JsonObject config);

    @Override
    public boolean isCapabilitySided(B buffer) {
        return false;
    }

    @Override
    public void attachCapabilities(ICapabilityInstanceConsumer target, B buffer) {
        // NO-OP
    }

    @Override
    public Unit createAccumulator() {
        return Unit.INSTANCE;
    }

    @Override
    public void accumulate(Unit acc, B buffer) {
        // NO-OP
    }

    @Override
    public Unit copyAccumulator(Unit acc) {
        return Unit.INSTANCE;
    }

    @Override
    public Unit createJeiAccumulator() {
        return Unit.INSTANCE;
    }

    @Override
    public void jeiAccumulate(Unit acc, Unit buffer) {
        // NO-OP
    }

    @Override
    public Collection<JeiUiElement<?>> createJeiUiElements(Unit buffer, ScreenRegion contRegion,
                                                           IJeiHelpers jeiHelpers) {
        return Collections.emptyList();
    }

    @FunctionalInterface
    protected interface VirtualBufferFactory<B> {

        B createBuffer(World world, BlockPos pos, @Nullable BufferObserver observer);

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/buffer/VirtualBufferType.kt
package st.evening.mc.cbtweaker.buffer

import mezz.jei.api.IJeiHelpers
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.compat.jei.ui.JeiUiElement
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson
import st.evening.mc.prelude.api.util.game.CapabilityVisitor
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.math.IntRectangle

abstract class VirtualBufferType<B> : BufferType<B, Unit, Unit, Unit> {
    override val accumulatorClass: Class<Unit>
        get() = Unit::class.java

    context(_: JsonPath)
    override fun loadBufferFactory(dto: TJson.Object): BufferFactory<B, Unit> {
        val bufFactory = loadVirtualBufferFactory(dto)
        return object : BufferFactory<B, Unit> {
            override fun createBuffer(world: World, pos: BlockPos, observer: BufferObserver): B =
                bufFactory.createBuffer(world, pos, observer)

            override fun createJeiBuffer(): Unit = Unit
        }
    }

    context(_: JsonPath)
    protected abstract fun loadVirtualBufferFactory(dto: TJson.Object): VirtualBufferFactory<B>

    override fun attachCapabilities(target: CapabilityVisitor, buffer: B) {}

    override fun createAccumulator(): Unit = Unit

    override fun accumulate(acc: Unit, buffer: B) {}

    override fun copyAccumulator(acc: Unit): Unit = Unit

    override fun createJeiAccumulator(): Unit = Unit

    override fun jeiAccumulate(acc: Unit, buffer: Unit) {}

    @ClientSide.Physical
    override fun createJeiUiElements(
        buffer: Unit,
        contRegion: IntRectangle,
        jeiHelpers: IJeiHelpers
    ): Collection<JeiUiElement<*>> = emptyList()

    protected fun interface VirtualBufferFactory<B> {
        fun createBuffer(world: World, pos: BlockPos, observer: BufferObserver): B
    }
}
